pipeline/runtime/step_executor.go

```diff
@@ -110,7 +113,122 @@ func (e *StepExecutor) PollStep(ctx context.Context, r *api.PollStepRequest) (*a
 	return convertStatus(status), nil
 }
 
+func (e *StepExecutor) StreamOutput(ctx context.Context, r *api.StreamOutputRequest) (oldOut []byte, newOut <-chan []byte, err error) {
+	id := r.ID
+	if id == "" {
+		err = &errors.BadRequestError{Msg: "ID needs to be set"}
+		return
+	}
+
+	var stepOutput *StepOutput
+
+	// the runner will call this function just before the call to start step, so we wait a while for the step to start
+	for ts := time.Now(); ; {
+		e.mu.Lock()
+		stepOutput = e.stepOutput[id]
+		e.mu.Unlock()
+
+		if stepOutput != nil {
+			break
+		}
+
+		const timeoutDelay = 5 * time.Second
+		if time.Since(ts) >= timeoutDelay {
+			err = &errors.BadRequestError{Msg: "Step has not started"}
+			return
+		}
+
+		const retryDelay = 100 * time.Millisecond
+		select {
+		case <-time.After(retryDelay):
+		case <-ctx.Done():
+			err = ctx.Err()
+			return
+		}
+	}
+
+	// subscribe to new data messages, and unsubscribe when the request context finished or when the step is done
+	chData := make(chan []byte)
+	oldOut, err = stepOutput.Subscribe(chData, r.Offset)
+	if err != nil {
+		return
+	}
+
+	go func() {
+		select {
+		case <-ctx.Done():
+		case <-stepOutput.Done():
+		}
+		close(chData)
+		stepOutput.Unsubscribe(chData)
+	}()
+
+	newOut = chData
```

Comment:
From this lines, it looks like all the data will be streamed after the step is complete.

Reply:
No. How could you conclude that? Quite the opposite. Streaming starts as soon as there is something to stream.